lib/_json.py

```diff
@@ -180,23 +180,22 @@ def _get_json_oids(conn_or_curs, name='json'):
     from psycopg2.extensions import STATUS_IN_TRANSACTION
     from psycopg2.extras import _solve_conn_curs
 
-    conn, curs = _solve_conn_curs(conn_or_curs)
-
-    # Store the transaction status of the connection to revert it after use
-    conn_status = conn.status
-
-    # column typarray not available before PG 8.3
-    typarray = conn.info.server_version >= 80300 and "typarray" or "NULL"
-
-    # get the oid for the hstore
-    curs.execute(
-        "SELECT t.oid, %s FROM pg_type t WHERE t.typname = %%s;"
-        % typarray, (name,))
-    r = curs.fetchone()
-
-    # revert the status of the connection as before the command
-    if conn_status != STATUS_IN_TRANSACTION and not conn.autocommit:
-        conn.rollback()
+    with _solve_conn_curs(conn_or_curs) as (conn, curs):
```

Comment:
If you use this syntax, don't you also start a context on the connection, forcing a final commit/rollback?

Reply:
No, it does not.  This enters the context manager `_solve_conn_curs()` which returns a tuple. This does not enter the context manager of the elements of the tuple. The context manager `_solve_conn_curs()`  also enters enters the cursor context manager, but not the one for the connection. Let me know if this is unclear and I can try to provide more details.